File: Two_pointers/Python/Longest Word in Dictionary through Deleting.py
# Problem link: https://leetcode.com/problems/longest-word-in-dictionary-through-deleting/description/
# Time: O(n * (m + k))
# Space: O(max(m, k))
class Solution:
    def findLongestWord(self, s: str, dictionary: List[str]) -> str:
        
        def isSub(word):
            idx = 0

            for c in s:
                if idx < len(word) and word[idx] == c:
                    idx += 1
            
            return idx == len(word)
        
        result = ""
        
        for word in dictionary:
            if isSub(word):
                if len(word) > len(result) or len(word) == len(result) and word < result:
                    result = word
        
        return result

File: Two_pointers/Java/Longest Word in Dictionary through Deleting.java
// Problem link: https://leetcode.com/problems/longest-word-in-dictionary-through-deleting/description/
// Time: O(n * (m + k))
// Space: O(max(m, k))
public class Solution {
    public String findLongestWord(String s, List<String> dictionary) {
        String result = "";

        for (String word : dictionary) {
            if (isSub(word, s)) {
                if (word.length() > result.length() || (word.length() == result.length() && word.compareTo(result) < 0)) {
                    result = word;
                }
            }
        }

        return result;
    }

    private boolean isSub(String word, String s) {
        int idx = 0;

        for (char c : s.toCharArray()) {
            if (idx < word.length() && word.charAt(idx) == c) {
                idx++;
            }
        }

        return idx == word.length();
    }
}
